File: livebox/src/main/java/com/creations/livebox/validator/AgeValidator.java
package com.creations.livebox.validator;

import com.creations.livebox.Livebox;
import com.creations.livebox.util.Optional;

import java.util.concurrent.TimeUnit;

import static java.util.concurrent.TimeUnit.DAYS;
import static java.util.concurrent.TimeUnit.HOURS;
import static java.util.concurrent.TimeUnit.MINUTES;

/**
 * @author Sérgio Serra on 01/09/2018.
 * Criations
 * sergioserra99@gmail.com
 */
public final class AgeValidator<R> implements Validator<R> {

    // Time to live before expiring the resources
    private long mTtl;
    private Journal mJournal;

    public static <T> AgeValidator<T> create(long ttl) {
        return new AgeValidator<>(Livebox.journal, ttl);
    }

    public static <T> AgeValidator<T> minutes(int minutes) {
        return create(MINUTES.toMillis(minutes));
    }

    public static <T> AgeValidator<T> hours(int hours) {
        return create(HOURS.toMillis(hours));
    }

    public static <T> AgeValidator<T> days(int days) {
        return create(DAYS.toMillis(days));
    }

    private AgeValidator(Journal journal, long ttl) {
        mJournal = journal;
        mTtl = ttl;
    }

    @Override
    public boolean validate(String key, R item) {
        Optional<Long> lastTimestampOpt = mJournal.read(key);
        // By default if no timestamp is available resource is valid
        return lastTimestampOpt.isAbsent() ||
                lastTimestampOpt.get() + mTtl >= System.currentTimeMillis();
    }
}

File: livebox/src/main/java/com/creations/livebox/validator/AgeValidator.kt
package com.creations.livebox.validator

import com.creations.livebox.Livebox
import java.util.concurrent.TimeUnit.*

/**
 * @author Sérgio Serra on 01/09/2018.
 * sergioserra99@gmail.com
 */
class AgeValidator<R> private constructor(
        private val mJournal: Journal?,
        // Time to live before expiring the resources
        private val mTtl: Long
) : Validator<R> {

    override fun validate(key: String, item: R): Boolean {

        if (mJournal == null) return true

        val lastTimestampOpt = mJournal.read(key)
        // By default if no timestamp is available resource is valid
        return lastTimestampOpt.isAbsent || lastTimestampOpt.get() + mTtl >= System.currentTimeMillis()
    }

    companion object {
        @JvmStatic
        fun <T> create(ttl: Long): AgeValidator<T> = AgeValidator(Livebox.journal, ttl)

        @JvmStatic
        fun <T> minutes(minutes: Int): AgeValidator<T> = create(MINUTES.toMillis(minutes.toLong()))

        @JvmStatic
        fun <T> hours(hours: Int): AgeValidator<T> = create(HOURS.toMillis(hours.toLong()))

        @JvmStatic
        fun <T> days(days: Int): AgeValidator<T> = create(DAYS.toMillis(days.toLong()))
    }
}

fun <T> Int.minutes(): AgeValidator<T> = AgeValidator.minutes(this)
fun <T> Int.hours(): AgeValidator<T> = AgeValidator.hours(this)
fun <T> Int.days(): AgeValidator<T> = AgeValidator.days(this)
